### DISCUSSION TITLE
Extension subtables in the repacker

### DISCUSSION THREAD

**Discussion by simoncozens** [harfbuzz/harfbuzz#2986] [Ideas] [answered]
I've been taking a look at the repacker code, and I have a thought. I'm wondering how you plan to handle moving lookups to extensions. Currently this is a todo item as part of handling overflow situations, but I wonder if this is the best approach. (I wonder because genuinely, I don't know.)

What I'm concerned about is when we promote things to extension lookups, that happens in a globally optimized rather than an individual way. (Yes, this *is* a hard problem.) For example, lookup subtables which share common descendants (particularly coverage tables) should be kept together.

I wonder if it would be better to think in terms of packing the lookups into sets of 64k pages, and optimising each page. Does that make sense?

**Comment by behdad** [comment, marked as answer]
Our repacker is good-enough to handle all objects together, instead of sharding them if they are promoted to Extension.

A very simple implementation would be to promote all lookups to Extension if we failed at all. Another idea is to write them all as Extension, then demote after packing if possible...

We somehow need to test that code first though. Currently we only subset fonts, which tends to not hit overflows easily.

**Reply by simoncozens** [reply]
So I wrote a *very* hacky utility, [`hb-repack`](https://gist.github.com/simoncozens/a36b5df8d2911685d0e64ed447c7cb1b), which takes a font, does a dummy subset to get it into a serializer, and then rams it through `hb_resolve_overflows` whether it needs it or not.

It's actually quite a fascinating test case for how well the repacker does.

Some results: a simple font with a GSUB table generated using fonttools.feaLib from this feature code:

```
lookup f_i_lig {sub f i by fi; } f_i_lig;
lookup s_x {sub s by x; } s_x;
feature rlig {
    sub f' lookup f_i_lig i' s' lookup s_x h';
} rlig;
```

Original (fonttools) GSUB size: 142 bytes.
After repacking: 148 bytes. (+6 bytes, 4%)

[Public Sans](https://github.com/uswds/public-sans) original (fonttools) GSUB table: 2108 bytes.
After repacking: 2192 bytes. (+84 bytes, 4%)

My crazy Nastaliq font killed the repacker, of course, but I didn't expect anything otherwise. But NotoNastaliqUrdu killed it as well: the repacked GPOS table was +7k more than fonttools, and then it gave up on GSUB.

So right now, I think it's fair to say (*if* my code is doing the right thing, which is dubious...) that the repacker does not generate the optimum layout, and that it struggles with anything with lots of overflows.

**Reply by behdad** [reply]
Excellent. That suggests that there's a bug somewhere in the repacker to begin with. cc @garretrieger

**Reply by garretrieger** [reply]
Thanks, this is a great idea for generating more difficult to repack test cases. The current implementation is still missing quite a bit so I'm not surprised that it's struggling with a more difficult case. Also the current implementation makes no attempt to reduce the size of the repacking. The primary goal is to just resolve offsets quickly.

If the size is being inflated excessively then my guess is that we are too aggressively duplicating shared tables and should maybe try re-ordering first. I'll take a look at your test cases and see where it's running into trouble which may help inform which parts should be implemented next.

**Reply by garretrieger** [reply]
I tried to replicate your findings and was unable. Here's what I did:

- Modified hb-subset.cc to always run the repacker (see: https://github.com/harfbuzz/harfbuzz/compare/main...googlefonts:always_repack?expand=1)
- Ran:

util/hb-subset --font-file=PublicSans-Regular.ttf --output-file=ps-subset.ttf --unicodes="*" --retain-gids --layout-features="*"

- Output was:

Glyph count: 649
Original size for GPOS = 13946
Repacked size for GPOS = 13946
Original size for GSUB = 2194
Repacked size for GSUB = 2194
Final Glyph count: 649

So I'm not seeing the increase after repacking that you are. I suspect that there's something you did incorrectly in setting up your hb-repack.cc. I haven't tested the nastaliq font yet, but I'll try that next.

**Reply by garretrieger** [reply]
I think where that increase may have come from was if you're comparing the unsubsetted GSUB tabled to the subsetted and repacked table. In which case I believe the increase is actually happening during subsetting and not repacking.

**Reply by garretrieger** [reply]
I have some ideas where those might be coming about. I believe we have a few cases where the serializer is not being used correctly and multiple sub tables get packed into a single serialize object which hampers efficient sub-table sharing. For example: https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-layout-gsub-table.hh#L104

This causes the coverage subtable to be serialized immediately after the end of the current subtable without doing a serializer.push() and pop_pack() which is where object deduplication happens.

**Reply by garretrieger** [reply]
Fix is here: https://github.com/harfbuzz/harfbuzz/pull/3020

**Reply by behdad** [reply]
Thanks Garret. Yes, that old pattern predated the push/pop_pack aka "tetris" serialization scheme. Thanks for debugging and cleaning up.

**Comment by garretrieger** [comment]
So I've made some major advances in the repackers handling of extension looks: https://github.com/harfbuzz/harfbuzz/pull/3248

With these changes the repacker will where possible keep extension lookup subgraphs packed together that are connected via shared nodes. This reduces unnecessary duplication and can significantly reduce the final packed size of GSUB/GPOS. This improves upon the fonttools approach of isolating each extension lookup subgraph.

The next step to this will be to enable extension lookup promotion which will allow non-extension lookups that have shared subtables with extension lookups to be automatically upgraded to extension lookups. Which will further reduce duplication.

**Reply by simoncozens** [reply]
This is looking really nice. Presumably the repacker will also demote extension lookups which are no longer necessary after subsetting?

One thing I don't see is any subtable sharing, where identical nodes can get the same pointer. Is that planned? (Or is it already in there and I haven't seen it?) You'll need that either when dealing with suboptimal fonts, or when you move a subtable from an extension to a non-extension (or vice versa) and it now has an opportunity to share its pointer that wasn't available before.

**Reply by garretrieger** [reply]
sub table sharing happens automatically during the initial serialization. The serializer detects when any subtable is packed that is identical to a previously packed sub table and points the offset to that instead (https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-serialize.hh#L292). So the repacker starts with a graph that has the maximum possible amount of sharing and then may selectively disable sharing of specific subtables/subgraphs via duplication as needed.
